File: nodejs/tests/motor/drv8830_test.js
'use strict';

const { I2CConnection } = require('../../packages/periph/src/connection/i2c');
const { DRV8830Minimal, DRV8830Full } = require('../../packages/periph/src/chips/motor/drv8830');

const I2C_BUS = parseInt(process.env.I2C_BUS || '1', 10);
const I2C_ADDR = parseInt(process.env.I2C_ADDR || '0x60', 16);

let passed = 0;
let failed = 0;

function checkTrue(label, condition) {
    if (condition) { console.log('PASS', label); passed++; }
    else { console.log('FAIL', label); failed++; }
}

async function main() {
    const connection = new I2CConnection(I2C_BUS, I2C_ADDR);
    const minimal = new DRV8830Minimal(connection);
    await minimal.init();
    checkTrue('init', true);

    const motor = new DRV8830Full(connection);
    await motor.drive(2.0);
    let out = await motor.readOutput();
    checkTrue('drive forward direction', out.direction === 'forward');
    checkTrue('drive forward voltage', Math.abs(out.voltage - 2.0) < 0.1);

    await motor.drive(-1.0);
    out = await motor.readOutput();
    checkTrue('drive reverse direction', out.direction === 'reverse');

    await motor.brake();
    checkTrue('brake direction', (await motor.readOutput()).direction === 'brake');

    await motor.stop();
    checkTrue('stop direction', (await motor.readOutput()).direction === 'coast');

    await motor.clearFault();
    checkTrue('clear fault', !(await motor.readFault()).fault);

    await connection.close();

    console.log(`===DONE: ${passed} passed, ${failed} failed===`);
    process.exit(failed === 0 ? 0 : 1);
}

main();

File: python/tests/motor/drv8830_test.py
"""MicroPython hardware test for the DRV8830 — runs against real hardware.

Drives the motor forward, reverse, brake and coast, reading back the CONTROL
register after each command, and checks the fault register can be cleared.
"""

import time
import machine
import _testconfig as cfg
from periph.connection.i2c_micropython import I2CConnection
from periph.chips.motor.drv8830 import DRV8830Minimal, DRV8830Full
from machine import Pin

passed = 0
failed = 0


def check_true(label, condition):
    global passed, failed
    if condition:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL', label)
        failed += 1


i2c = machine.I2C(cfg.I2C_ID, sda=Pin(cfg.SDA), scl=Pin(cfg.SCL), freq=cfg.FREQ)
connection = I2CConnection(i2c, cfg.ADDR)

motor = DRV8830Minimal(connection)
check_true('construct_minimal', isinstance(motor, DRV8830Minimal))

full = DRV8830Full(connection)
full.drive(2.0)
time.sleep(0.1)
voltage, direction = full.read_output()
check_true('drive_forward_direction', direction == 'forward')
check_true('drive_forward_voltage', abs(voltage - 2.0) < 0.1)

full.drive(-1.0)
voltage, direction = full.read_output()
check_true('drive_reverse_direction', direction == 'reverse')

full.brake()
check_true('brake_direction', full.read_output()[1] == 'brake')

full.stop()
check_true('stop_direction', full.read_output()[1] == 'coast')

full.clear_fault()
check_true('clear_fault', not full.read_fault()[0])

print('===DONE: %d passed, %d failed===' % (passed, failed))
